src/fstests/fstests.jinja2

```diff
@@ -16,35 +23,87 @@ class Job(BaseJob):
             result = subprocess.run("kvm-xfstests install-kconfig", check=True, shell=True)
             if result.returncode == 0:
                 print(f"Kernel config done.")
-                return 'pass'
+                return True
             else:
                 print((f"Missing kernel .config file."))
-                return 'fail'
+                return False
         except Exception as e:
             print('Kernel config error:', e)
-            return 'fail'
+            return False
 
     def _kernel_build(self, kdir):
         try:
             os.chdir(kdir)
             result = subprocess.run("make -j$(nproc)", check=True, shell=True)
             if result.returncode == 0:
                 print(f"Kernel Build DONE.")
-                return 'pass'
+                return True
             else:
-                return 'fail'
+                return False
         except Exception as e:
             print('Kernel build error:', e)
 
+    def _get_xml_from_vm(self, kdir):
+        try:
+            result = subprocess.run([f"debugfs -R \"dump ext4/results-4k/results.xml {kdir}/results.xml\" {xfstests_bld_path}/run-fstests/disks/vdg"], check=True, shell=True)
+            if result.returncode == 0:
+                print(f"XML file in the output directory.")
+                return True
+        except Exception as e:
+            print('XML not found.', e)
+            return False
+
+    def _parse_xml_results(self, kdir):
+        xml_file = ET.parse(kdir)
+        root = xml_file.getroot()
+        tree = {}
+        for test_case in root.iter('testcase'):
+            if test_case.find('skipped') is not None:
+                result = None
+            elif test_case.find('failure') is not None:
+                result = 'fail'
+            else:
+                result = 'pass'
+            name = test_case.attrib['name'].split('/')
+            if not name[0] == 'ext4':
+                name.insert(0, 'ext4')
+            current = tree
+            for item in name:
+                if not item in current:
+                    current[item] = {}
+                current = current[item]
+            current['result'] = result
+        return tree
+
     def _run_smoke_tests(self):
         try:
             result = subprocess.run(["kvm-xfstests", "smoke"], check=True)
             if result.returncode == 0:
                 print(f"Test run success.")
-                return 'pass'
+                return True
         except Exception as e:
             print('Problem found during tests:', e)
-            return 'fail'
+            return False
+
+    def _convert_results_for_api(self, tree, name):
+        if 'result' in tree:
+            return {
+                'name': name,
+                'result': tree['result'],
+            }
+        else:
+            child_nodes = []
+            for child_name in tree:
+                child_nodes.append(self._convert_results_for_api(tree[child_name], child_name))
```

Comment:
This is a very long line too.  It could easily be wrapped, or an intermediate variable with the results could then be appended to the list.